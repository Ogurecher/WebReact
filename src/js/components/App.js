import React, { Component } from "react";
import GeolocationButton from "./GeolocationButton.js"
import InputForm from "./InputForm.js"
import Favourites from "./Favourites.js";
import CityInfo from './CityInfo';
import { connect } from "react-redux";
import { getWeather } from "../actions/index";

export class App extends Component {

  componentDidMount() {
    //add getweather for all cities
    for (let i = 0; i < this.props.cities.length; i++) {
      this.props.getWeather(this.props.cities[i].position, i);
    }
  }

  componentDidUpdate(prevProps) {
    if (JSON.stringify(this.props.position) !== JSON.stringify(prevProps.position)) {
      this.props.getWeather(this.props.position, null);
    }
    //add getweather for all cities
    for (let i = 0; i < this.props.cities.length; i++) {
      if (!prevProps.cities[i] || JSON.stringify(this.props.cities[i].weather) !== JSON.stringify(prevProps.cities[i].weather)) {  //may cause problems if cities[i] does not exist in prevprops
        this.props.getWeather(this.props.cities[i].position, i);
      }
    }
  }

  render() {
    return (
      <div>
        <GeolocationButton />
        <CityInfo city={{weather: this.props.weather, loading: this.props.loading, errorMsg: this.props.errorMsg}} />
        <InputForm />
        <Favourites cities={this.props.cities} />
      </div>
    );
  }
}


function mapStateToProps(state) {
  return {
    position: state.position,
    weather: state.weather,
    loading: state.loading,
    errorMsg: state.errorMsg,
    cities: state.cities
  };
}

export default connect(
  mapStateToProps,
  { getWeather }
)(App);