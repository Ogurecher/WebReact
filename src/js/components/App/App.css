body {
    font-family: Verdana;
    color: darkslategrey;
}