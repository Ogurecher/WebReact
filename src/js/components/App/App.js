import React, { Component } from 'react';
import GeolocationButton from '../GeolocationButton/GeolocationButton'
import InputForm from '../InputForm/InputForm'
import Favourites from '../Favourites/Favourites';
import CityInfo from '../CityInfo/CityInfo';
import { connect } from 'react-redux';
import { getWeather } from '../../actions/index';
import './App.css';

export class App extends Component {

  componentDidMount() {
    for (let i = 0; i < this.props.cities.length; i++) {
      this.props.getWeather(this.props.cities[i].position, i);
    }
  }

  componentDidUpdate(prevProps) {
    if (JSON.stringify(this.props.position) !== JSON.stringify(prevProps.position)) {
      this.props.getWeather(this.props.position, null);
    }

    for (let i = 0; i < this.props.cities.length; i++) {
      if (!prevProps.cities[i] || JSON.stringify(this.props.cities[i].weather) !== JSON.stringify(prevProps.cities[i].weather)) {
        this.props.getWeather(this.props.cities[i].position, i);
      }
    }
  }

  render() {
    return (
      <div>
        <GeolocationButton />
        <CityInfo city={{weather: this.props.weather, loading: this.props.loading, errorMsg: this.props.errorMsg}} id='main' />
        <InputForm />
        <Favourites cities={this.props.cities} favourites={this.props.favourites} />
      </div>
    );
  }
}


function mapStateToProps(state) {
  return {
    position: state.position,
    weather: state.weather,
    loading: state.loading,
    errorMsg: state.errorMsg,
    cities: state.cities,
    favourites: state.favourites
  };
}

export default connect(
  mapStateToProps,
  { getWeather }
)(App);