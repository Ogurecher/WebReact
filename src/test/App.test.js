import React from 'react';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';

import App from '../js/components/App/App';

const middlewares = [thunk];
const store = configureMockStore(middlewares);

describe('App', function() {
    test('snapshot renders', function() {
        const storeMock = store({
            cities: [
                {
                    position: {
                        city: 'London'
                    }
                }
            ],
            favourites: ['London']
        });

        const geolocationMock = {
            getCurrentPosition: jest.fn()
        };
        global.navigator.geolocation = geolocationMock;

        const weatherMock = {};

        global.fetch = () => {
            return Promise.resolve(weatherMock);
        };

        const component = renderer.create(<Provider store={storeMock} ><App /></Provider>);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});