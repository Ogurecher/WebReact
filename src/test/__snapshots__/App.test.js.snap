// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot renders 1`] = `
<div>
  <div
    className="geolocation-wrapper"
  >
    <p
      className="geolocation-text"
    >
      Weather here
    </p>
    <form
      onSubmit={[Function]}
    >
      <input
        type="submit"
        value="Get location"
      />
    </form>
  </div>
  <p />
  <form
    className="favourites-form-wrapper"
    onSubmit={[Function]}
  >
    <p
      className="favourites-text"
    >
      Favourites
    </p>
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
    <input
      type="submit"
      value="+"
    />
  </form>
  <div
    className="favourites-cities-wrapper"
  >
    <p />
  </div>
</div>
`;
